### PR TITLE
Entity visibility control

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR introduces `VisiblityGuard` which allows to get an access to a registered repository depending on the visibility state of entities this repository manages. 

`BoundedContext` now is able to return type names of entity states by their visibility.

In addition:
* `Repository.isOpen()` made public.

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@armiol, PTAL.

File: client/src/main/java/org/spine3/option/EntityOptions.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * @author Alexander Yevsyukov
+ */
+@Internal
+public class EntityOptions {
+
+    private EntityOptions() {
+        // Prevent instantiation of this utility class.
+    }
+
+    public static Visibility getVisibility(Class<? extends Message> stateClass) {
```
Comment (armiol) [inline client/src/main/java/org/spine3/option/EntityOptions.java]: Let's check the parameter for `null` and test this case via `NullPointerTester`.

File: server/src/main/java/org/spine3/server/BoundedContext.java [hunk 2]
```diff
@@ -316,39 +292,55 @@ public void notify(IntegrationEvent integrationEvent,
...
      * @param aggregateStateClass the class of the aggregate state
      * @return repository instance or empty {@code Optional} if not found
      */
     public Optional<? extends AggregateRepository<?, ?>> getAggregateRepository(
             Class<? extends Message> aggregateStateClass) {
-        final AggregateRepository<?, ?> result = aggregateRepositories.get(aggregateStateClass);
-        return Optional.fromNullable(result);
+        // See if there is a repository for this state at all.
+        if (!guard.hasRepository(aggregateStateClass)) {
+            throw newIllegalStateException("No repository found for the aggregate state class %s",
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/BoundedContext.java]: This branch is reported to be untested.

File: server/src/main/java/org/spine3/server/entity/VisibilityGuard.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+     *
+     * @param stateClass the class of the state of entities managed by the repository
+     * @return the repository wrapped into {@code Optional} or {@code Optional#absent()} if the
+     * entity state is {@linkplain Visibility#NONE not visible}
+     * @throws IllegalArgumentException if the repository for the passed state class was not
+     *                                  {@linkplain #register(Repository) registered} with the guard
+     *                                  prior to this call, or if all repositories were
+     *                                  {@linkplain #shutDownRepositories() shut down}
+     */
+    public Optional<Repository> getRepository(Class<? extends Message> stateClass) {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/VisibilityGuard.java]: As long we we perform the `checkNotNull(...)` in `getEntityTypes(visibility)` below, I think we should do the same for `stateClass` as well.

File: server/src/main/java/org/spine3/server/entity/VisibilityGuard.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+ * A registry of repositories that controls access to them depending on the visibility of
+ * corresponding entity states.
+ *
+ * @author Alexander Yevsyukov
+ * @see EntityOptions
+ */
+@Internal
+public final class VisibilityGuard {
+
+    private final Map<Class, RepositoryAccess> repositories = newHashMap();
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/VisibilityGuard.java]: I would make the bounds more strict: `Map<Class<? extends Message>, RepositoryAccess>`.

File: server/src/main/java/org/spine3/server/entity/VisibilityGuard.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+     * Registers the passed repository with the guard.
+     */
+    public void register(Repository repository) {
+        checkNotNull(repository);
+        final Class stateClass = repository.getEntityStateClass();
+        checkNotAlreadyRegistered(stateClass);
+        repositories.put(stateClass, new RepositoryAccess(repository));
+    }
+
+    private void checkNotAlreadyRegistered(Class stateClass) {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/VisibilityGuard.java]: Let's check this parameter for `null` in order to be consistent. 

See more below on the same matter.

File: server/src/test/java/org/spine3/server/entity/VisibilityGuardShould.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+public class VisibilityGuardShould {
+
+    private VisibilityGuard guard;
+    private List<Repository> repositories;
+    private BoundedContext boundedContext;
+
+    @Before
+    public void setUp() {
+        boundedContext = BoundedContext.newBuilder()
+                                                            .build();
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/entity/VisibilityGuardShould.java]: Please fix the wrapping.

File: server/src/main/java/org/spine3/server/BoundedContext.java (hunk 2 above)
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/BoundedContext.java]: Seems to be a bug of Codecov. There's a test for it. Please see `BoundedContextShould.throw_ISE_when_no_repository_registered()`.

File: client/src/main/java/org/spine3/option/EntityOptions.java (hunk 1 above)
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/option/EntityOptions.java]: Done

File: server/src/main/java/org/spine3/server/entity/VisibilityGuard.java (hunk 3 above)
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/entity/VisibilityGuard.java]: Done

**Comment by alexander-yevsyukov** [comment]
@armiol, PTAL.

**Pull request merged**